Adding new, more powerful search

henryiii opened the pull request:
More Pythonic search (the original function was designed around the orignal Pandas table). Some examples of usage:

```python
# These return [p, pbar]
Particle.from_search_list(lambda p: p.name == 'p')
Particle.from_search_list(name = 'p')
Particle.from_search_list(name = lambda x: x == 'p')

# These return 1 result, [p]
Particle.from_search_list(lambda x: x.name == 'p' and x > 0)
Particle.from_search_list(name = 'p', particle=True)

# These return all particles with `p` in the fullname
Particle.from_search_list('p')
Particle.from_search_list(fullname = lambda x: 'p' in x)

# Comparisons, etc.
Particle.from_search_list(lifetime = lambda x : x > 1*second) 
```

eduardo-rodrigues commented:
I like the idea that one can still do simple things as before, such as `Particle.from_search_list(name = 'p')`, while being exposed to a lot of flexibility and power with for example `Particle.from_search_list(lifetime = lambda x : x > 100000)`.

I did not understand
`Particle.from_search_list(name = 'p', particle=True) # exact match, 1 results`

since the from_search_list is there to return a list of matches.

I look forward to some tests,, which will also serve as examples. Worth updating the README too.

Thanks!

henryiii commented:
> since the from_search_list is there to return a list of matches.

I'm just adding a note there about what it returns. It's still a list. For that search, it returns a proton. The one(s) above return protons and antiprotons.

henryiii commented:
I've added tests, fixed an existing bug with the location of `*`, and added better minimum requirements (Requested by @HDembinski). I've dropped the special regular expression handling; I assume any user familiar with regex's could also write a lambda function. This should be good to go unless someone has feedback.

I don't allow unlimited lambda functions with `and` style combining; it's a little uglier to do right in Python 2. I think the current system is versatile enough.

The meaning of an empty search may have changed, and this is an API change for this function.

henryiii commented:
Note: Are we planning to release a 0.2.1 before 0.3.0? A few of the fixes need to go into 0.2.1 if so, but the API change needs to be only in 0.3.0.

eduardo-rodrigues commented:
That was the plan, see https://github.com/scikit-hep/particle/milestone/1. But we can go straight for 0.3.0 if that's easiest at this stage. Otherwise, as you say, this PR has things for 0.2.1 and others for 0.3.0, it seems.
What's your preference?

eduardo-rodrigues commented on README.rst:
```diff
@@ -127,7 +127,7 @@ can use a search:
     >>> Particle.from_pdgid(211)
     <Particle: pdgid=211, fullname='pi+', mass=139.57061 ± 0.00024 MeV>
     >>>
-    >>> Particle.from_search_list('pi')[0]
```
By removing this line you mean that this simple shortcut with just a string as name search won't be possible anymore? Not the end of the world but a bit of a shame IMO.

henryiii replied:
This has become `name='pi'`, but that's exact match only. So the proper replacement for this is:

```Python
Particle.from_search_list(name = lambda x: 'pi' in x)[0]
```

Would it make sense to use the first argument as a fuzzy name search if it is not callable? Then this shortcut would return.

eduardo-rodrigues replied:
I'm confused now since from_search_list is different from from_search in that the former returns a list of matches. I guess the answer to your question is yes - if I understand correctly.

eduardo-rodrigues replied:
Thanks for that.

eduardo-rodrigues commented:
Let me know about the idea of getting this all into release 0.2.1 or not … Thanks.

henryiii commented:
I pulled out the fixes for 0.2.1. We can try to push a release very soon, then merge this toward 0.3.0.

eduardo-rodrigues commented:
OK, thanks.

eduardo-rodrigues commented:
I will tag a release 0.2.1 later today unless I hear from you saying to wait …

eduardo-rodrigues commented:
Version 0.2.1 is tagged and deployed. You can now rebase ...

henryiii commented:
Any ideas why Azure didn't try to run?

eduardo-rodrigues commented:
I've noticed Azure do that every now and then but have no idea why it happens. When it happens, I go to the Checks tab and "re-run all".

eduardo-rodrigues commented:
This is good to go, @henryiii?

eduardo-rodrigues commented:
FYI I've just released and deployed version 0.2.2 to fix a bug reported by Hans. You will need to pull in the updates after you merge this.
